README.md
# pasta-curves

[![Stack CI](https://github.com/nccgroup/pasta-curves/actions/workflows/stack.yml/badge.svg)](https://github.com/nccgroup/pasta-curves/actions/workflows/stack.yml)
[![Cabal CI](https://github.com/nccgroup/pasta-curves/actions/workflows/cabal.yml/badge.svg)](https://github.com/nccgroup/pasta-curves/actions/workflows/cabal.yml)
[![Hackage](https://img.shields.io/hackage/v/pasta-curves.svg?logo=haskell)](https://hackage.haskell.org/package/pasta-curves)
[![Stackage Lts](http://stackage.org/package/pasta-curves/badge/lts)](http://stackage.org/lts/package/pasta-curves)
[![Stackage Nightly](http://stackage.org/package/pasta-curves/badge/nightly)](http://stackage.org/nightly/package/pasta-curves)
[![MIT license](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)


This Haskell library provides the Pasta Curves consisting of: the `Pallas`
curve and its `Fp`  field element, the `Vesta` curve  and its `Fq` field 
element, and a variety of  supporting functionality such as point/element 
arithmetic, serialization, and hash-to-curve. The algorithms are NOT constant
time.

Pallas is y<sup>2</sup> = x<sup>3</sup> + 5 over F<sub>p</sub>(0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001).
The order of the Pallas curve is 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001.


Vesta is y<sup>2</sup> = x<sup>3</sup> + 5 over F<sub>q</sub>(0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001).
The order of the Vesta curve is 0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001.

The curves are designed such that the order of one matches the field 
characteristic of the other. For a brief introduction, see the Zcash blog 
titled ["The Pasta Curves for Halo 2 and Beyond"](https://electriccoin.co/blog/the-pasta-curves-for-halo-2-and-beyond/).
The reference Rust implementation (which inspired this implementation) 
can be found at: <https://github.com/zcash/pasta_curves>.

Example usage of this library implementation:

~~~
$ cabal repl

ghci> a = 9 :: Fp

ghci> a*a
0x0000000000000000000000000000000000000000000000000000000000000051

ghci> pointMul a base :: Vesta
Projective {_px = 0x3CDC6A090F2BB3B52714C083929B620FE24ADBCBBD420752108CD7C29E543E5E, 
            _py = 0x08795CD330B3CE5AA63BD2B18DE155AE3C96E8AF9DA2CC742C6BA1464E490161, 
            _pz = 0x1FA26F58F3A641ADFE81775D3D53378D6178B6CCBF14F9BD4AB5F10DEE28D878}
~~~

---

Copyright 2022 Eric Schorn; Licensed under the MIT License.

CHANGELOG.md
# Changelog

`pasta-curves` uses [PVP Versioning][1].
The changelog is available [on GitHub][2].

## 0.0.0.0

* Initially created.

## 0.0.1.0

* Dropped (minimum required) cabal version to improve compatibility
* Added functions to generate A) invertible random field element, and B) random Pallas/Vesta curve point
* Added `Bounded` instances to field elements


[1]: https://pvp.haskell.org
[2]: https://github.com/nccgroup/pasta-curves/releases


test/Spec.hs
{-# LANGUAGE NoImplicitPrelude, Trustworthy #-}

module Main (main) where

import Prelude
import System.Environment (setEnv)
import Test.Tasty (defaultMain, testGroup)
import TestFields (fieldProps, testBadF, testGoodF, testRnd)
import TestCurves (curveProps, testHashToPallas, testHashToVesta, testPOI, testBadC, testPallasEq, testRndPV)


main :: IO ()
main = do
  setEnv "TASTY_QUICKCHECK_TESTS" "1_000"
  defaultMain $ testGroup "\nRunning Tests" [fieldProps, testBadF, testGoodF, testRnd,
    testHashToPallas, testHashToVesta, curveProps, testPOI, testBadC, testPallasEq, testRndPV]
  print "Finished!"


test/TestCurves.hs
-- Test curves...

{-# LANGUAGE FlexibleInstances, NoImplicitPrelude, Trustworthy #-}
{-# OPTIONS_GHC -Wno-orphans #-}

module TestCurves (curveProps, testPOI, testHashToPallas, testHashToVesta, 
  testBadC, testPallasEq, testRndPV) where

import Prelude hiding (exp)
import Data.ByteString (pack)
import Data.ByteString.UTF8 (fromString)
import Data.Maybe (fromJust, isNothing)
import Data.Word (Word8)
import System.Random (mkStdGen, StdGen)
import Test.Tasty (TestTree, testGroup)
import Test.Tasty.HUnit (assertBool, testCase)
import Test.Tasty.QuickCheck (Arbitrary(..), testProperty)
import TestFields ()
import PastaCurves
import Curves (Point(Projective))


instance Arbitrary Pallas where
  arbitrary = do
    scalar <- arbitrary
    return $ pointMul (scalar :: Fq) (base :: Pallas)  


instance Arbitrary Vesta where
  arbitrary = do
    scalar <- arbitrary
    return $ pointMul (scalar :: Fp) (base :: Vesta)  


curveProps :: TestTree
curveProps = testGroup "Testing Curve properties via QuickCheck" [

  testProperty "Pallas point add/mul" $ 
    \x y -> pointAdd (pointMul (x :: Fq) base) (pointMul y base) == pointMul (x+y) (base::Pallas),
  testProperty "Pallas point add symm" $
    \x y -> pointAdd x y == pointAdd y (x :: Pallas),
  testProperty "Pallas ser->deser" $
    \x -> fromJust (fromBytesC (toBytesC x)) == (x :: Pallas),

  testProperty "Vesta point add/mul" $ 
    \x y -> pointAdd (pointMul (x :: Fp) base) (pointMul y base) == pointMul (x+y) (base :: Vesta),
  testProperty "Vesta point add symm" $
    \x y -> pointAdd x y == pointAdd y (x :: Vesta),
  testProperty "Vesta ser->deser" $
    \x -> fromJust (fromBytesC (toBytesC x)) == (x :: Vesta)
  ] 


testPOI :: TestTree
testPOI = testCase "poi decode" $
  do
    let poiBytes = pack [0]
    let orderLessOne = (0::Fq) - 1
    let act = pointAdd (pointMul orderLessOne base) base :: Pallas
    let actBytes = toBytesC act
    assertBool "bad pairing mul" (poiBytes == actBytes)
    assertBool "point point eq" (act == (neutral :: Pallas))


testBadC :: TestTree
testBadC = testCase "bad decode" $ do
  let tooFewBytes = pack [0, 0]
  let act1 = fromBytesC tooFewBytes :: Maybe Pallas
  assertBool "bad too short bytes" (isNothing act1)
  let tooManyBytes = pack $ replicate 34 0
  let act2 = fromBytesC tooManyBytes :: Maybe Pallas
  assertBool "bad too many bytes" (isNothing act2)
  let tooLargeValue = pack $ (0x02 :: Word8) : (0x41 :: Word8) : replicate 31 0
  let act3 = fromBytesC tooLargeValue :: Maybe Pallas
  assertBool "bad too large value" (isNothing act3)
  let notOnCurve = pack $ (0x02 :: Word8) : replicate 32 0
  let act4 = fromBytesC notOnCurve :: Maybe Pallas
  assertBool "bad not on curve" (isNothing act4)
  let baseBytes = pack $ ((0x02 :: Word8) : replicate 31 0) ++ [0x01 :: Word8]
  let act5 = fromBytesC baseBytes :: Maybe Pallas
  assertBool "bad is base" $ fromJust act5 == (base :: Pallas)
  let notBaseBytes = pack $ ((0x03 :: Word8) : replicate 31 0) ++ [0x01 :: Word8]
  let act6 = fromBytesC notBaseBytes :: Maybe Pallas
  assertBool "bad not base" $ fromJust act6 /= (base :: Pallas)


testPallasEq :: TestTree
testPallasEq = testCase "bad Eq" $ do
  let x1 = 0x2e341f9b583ed433336de60408dc32487b37d8076f09ed1cd25e4f406c46f0bc :: Fp
  let y1 = 0x285a83a7d3a0d903aa1bb19eb6894d7dca04b13687abf05423ce54362de3d5de :: Fp
  let z1 = 1 :: Fp
  let x2 = 0x24024b8e1f42b83c83c211405147f74209b30aca025efc773734e7163798eb77 :: Fp
  let y2 = 0x27db4a65be914fba5eac9c613554e58b4469b8eb344e937511f48b1780d9ab70 :: Fp
  let z2 = 1 :: Fp
  let x3 = 0x1dc72099914e05a28ce349b110c6f52291eb1cec24643bbed1fb489f7cb41f47 :: Fp
  let y3 = 0x01b1f3ee3ec752c0a9022c1c2e178d00c10aa97417236f91a49ab232ec347c4f :: Fp
  let z3 = 1 :: Fp
  assertBool "Bad Eq1" $ (Projective x1 y1 z1 :: Pallas) /= Projective x2 y2 z2
  assertBool "Bad Eq2" $ (Projective x2 y2 z2 :: Pallas) /= Projective x3 y3 z3
  assertBool "Bad Eq3" $ (Projective x3 y3 z3 :: Pallas) /= Projective x1 y1 z1


testHashToPallas :: TestTree
testHashToPallas = testCase "testHashToPallas" $ assertBool "Failed testHashToPallas" helper
  where
    actual = hashToPallas (fromString "Trans rights now!")  -- String from zcash test vector line 147 (link below)
    -- See https://github.com/zcash/pasta_curves/blob/21fd9e2c1bbd2d049bfe95588d77cb884e9f93ab/src/pallas.rs#L150-L158
    z = 0x1d48103df8fcbb70d1809c1806c95651dd884a559fec0549658537ce9d94bed9 :: Fp
    x = 0x36a6e3a9c50b7b6540cb002c977c82f37f8a875fb51eb35327ee1452e6ce7947 * inv0 (z ^ (2::Integer))
    y = 0x01da3b4403d73252f2d7e9c19bc23dc6a080f2d02f8262fca4f7e3d756ac6a7c * inv0 (z ^ (3::Integer))
    expected = Projective x y 1 :: Pallas    
    helper = actual == expected


testHashToVesta :: TestTree
testHashToVesta = testCase "testHashToVesta" $ assertBool "Failed testHashToVesta" helper
  where
    actual = hashToVesta (fromString "hello")  -- String from zcash test vector line 147 (link below)
    -- See https://github.com/zcash/pasta_curves/blob/21fd9e2c1bbd2d049bfe95588d77cb884e9f93ab/src/vesta.rs#L63-L71
    z = 0x1b58d4aa4d68c3f4d9916b77c79ff9911597a27f2ee46244e98eb9615172d2ad :: Fq
    x = 0x12763505036e0e1a6684b7a7d8d5afb7378cc2b191a95e34f44824a06fcbd08e * inv0 (z ^ (2::Integer))
    y = 0x0256eafc0188b79bfa7c4b2b393893ddc298e90da500fa4a9aee17c2ea4240e6 * inv0 (z ^ (3::Integer))
    expected = Projective x y 1 :: Vesta
    helper = actual == expected

testRndPV :: TestTree
testRndPV = testCase "testRnd" $
  do
    let (r1, f1) = rndPallas (mkStdGen 1) :: (StdGen, Pallas)
    let (r2, f2) = rndPallas r1 :: (StdGen, Pallas)
    assertBool "testRndP" (pointAdd f1 f2 == pointAdd f2 f1)
    let (r3, f3) = rndVesta r2 :: (StdGen, Vesta)
    let (_r4, f4) = rndVesta r3 :: (StdGen, Vesta)
    assertBool "testRndV" (pointAdd f3 f4 == pointAdd f4 f3)


test/TestFields.hs
-- Test fields...

{-# LANGUAGE DataKinds, FlexibleInstances, NoImplicitPrelude, OverloadedStrings, Trustworthy #-}
{-# OPTIONS_GHC -Wno-orphans -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Redundant bracket" #-}

module TestFields (fieldProps, testBadF, testGoodF, testRnd) where

import Prelude hiding (sqrt)
import Data.ByteString (pack)
import Data.Maybe (fromJust, isNothing)
import Data.Word (Word8)
import System.Random (mkStdGen, StdGen)
import Test.Tasty (TestTree, testGroup)
import Test.Tasty.HUnit (testCase, assertBool)
import Test.Tasty.QuickCheck (Arbitrary(..), choose, testProperty)
import PastaCurves


instance Arbitrary Fp where
   arbitrary = fromInteger <$> choose (0, 2 ^ (512::Integer))

instance Arbitrary Fq where
   arbitrary = fromInteger <$> choose (0, 2 ^ (512::Integer))


fieldProps :: TestTree
fieldProps = testGroup "Testing Field properties via QuickCheck" [
  testProperty "Fp arith"  $ \a b c -> a*(b-c) - a*b + a*c == (0 :: Fp),
  testProperty "Fp inv0"   $ \a -> a * inv0 a == (1 :: Fp),
  testProperty "Fp sqrt"   $ \a -> fromJust (sqrt (a*a)) ^ (2 :: Integer) == (a ^ (2 :: Integer) :: Fp),
  testProperty "Fp isSqr"  $ \a -> isSqr (a*a :: Fp),
  testProperty "Fp serdes" $ \a -> fromBytesF (toBytesF a) == Just (a :: Fp),
  testProperty "Fp shiftR1" $ \a -> shiftR1 a == fromInteger (toI (a :: Fp) `div` 2),

  testProperty "Fq arith"  $ \a b c -> a*(b-c) - a*b + a*c == (0 :: Fq),
  testProperty "Fq inv0"   $ \a -> a * inv0 a == (1 :: Fq),
  testProperty "Fq sqrt"   $ \a -> fromJust (sqrt (a*a)) ^ (2 :: Integer) == (a ^ (2 :: Integer)  :: Fq),
  testProperty "Fq isSqr"  $ \a -> isSqr (a*a :: Fq),
  testProperty "Fq serdes" $ \a -> fromBytesF (toBytesF a) == Just (a :: Fq),
  testProperty "Fq shiftR1" $ \a -> shiftR1 a == fromInteger (toI (a :: Fq) `div` 2)
  ]


testBadF :: TestTree
testBadF = testCase "testBadF" $
  do
    let tooFewBytes = pack [0]
    let act1 = fromBytesF tooFewBytes :: Maybe Fq
    assertBool "bad too short bytes" (isNothing act1)
    let tooManyBytes = pack $ replicate 33 0
    let act2 = fromBytesF tooManyBytes :: Maybe Fq
    assertBool "bad too many bytes" (isNothing act2)
    let tooLargeValue = pack $ (0x41 :: Word8) : replicate 31 0
    let act3 = fromBytesF tooLargeValue :: Maybe Fq
    assertBool "bad too large value" (isNothing act3)

testGoodF :: TestTree
testGoodF = testCase "testGoodF" $
  do
    assertBool "sgn0 -1" (sgn0 (negate (1 :: Fq)) == 0)
    assertBool "sgn0 1" (sgn0 (1 :: Fq) == 1)
    -- hash2Field tested as part of hash2Curve

testRnd :: TestTree
testRnd = testCase "testRnd" $
  do
    let (r1, f1) = rndF (mkStdGen 1) :: (StdGen, Fp)
    let (_r2, f2) = rndF r1 :: (StdGen, Fp)
    assertBool "testRnd" (f1 + f2 == f2 + f1)
    -- hash2Field tested as part of hash2Curve

benchmark/Main.hs
{-|
Module      : benchmark.Main (internal)
Description : Target for benchmarking build.
Copyright   : (c) Eric Schorn, 2022
Maintainer  : eric.schorn@nccgroup.com
Stability   : experimental
Portability : GHC
SPDX-License-Identifier: MIT

This module provides a basic Main target for benchmarking builds. The suite of
benchmarks will be expanded over time.
-}

{-# LANGUAGE NoImplicitPrelude, Safe #-}

module Main (main) where

import Prelude
import Data.ByteString.UTF8 (fromString)
import Criterion.Main
import PastaCurves
import Criterion.Types (timeLimit, resamples)


benchConfig = defaultConfig {
              timeLimit = 60,
              resamples = 20
           }

-- Benchmarks will be further developed in future versions...
main :: IO ()
main = defaultMainWith benchConfig [
  bgroup "Group 1" [ 
    bench "Hello Pallas 1"  $ whnf hashToPallas (fromString "Hello Pallas 1"), 
    bench "Hello Pallas 2"  $ whnf hashToPallas (fromString "Hello Pallas 2"), 
    bench "Hello Vesta 1"   $ whnf hashToVesta  (fromString "Hello Vesta 1"), 
    bench "Hello Vesta 2"   $ whnf hashToVesta  (fromString "Hello Vesta 2"), 
    bench "Pasta base * -1" $ whnf (pointMul (-1 :: Fq)) (base :: Pallas),
    bench "Pasta base * 2^200 - 1" $ whnf (pointMul (2 ^ (200::Integer) - 1 :: Fq)) (base :: Pallas),
    bench "Pasta base * 2^100 - 1" $ whnf (pointMul (2 ^ (100::Integer) - 1 :: Fq)) (base :: Pallas),
    bench "Vesta base * -1" $ whnf (pointMul (-1 :: Fp)) (base :: Vesta)
    ]
  ]

-- Note: Initial field multiply performance is in line with expected Rust BigInt described at
-- https://research.nccgroup.com/2021/09/10/optimizing-pairing-based-cryptography-montgomery-multiplication-in-assembly/

{- Potential performance improvements to investigate/implement
0. Inline (<- almost zero) and other compiler directives (-O2 <- almost zero)
1. Remove field add/sub MOD by compare then subtract (<- almost zero)
2. Remove field mul MOD by Barrett reduction / Montgomery multiplication
3. Implement explicit pointDouble routine
4. Separate out statics (e.g. in sqrt routines)

Base Case
=========
benchmarking Group 1/Hello Pallas 1
time                 2.213 ms   (2.190 ms .. 2.257 ms)

benchmarking Group 1/Hello Pallas 2
time                 2.437 ms   (2.428 ms .. 2.441 ms)

benchmarking Group 1/Hello Vesta 1
time                 2.458 ms   (2.454 ms .. 2.464 ms)

benchmarking Group 1/Hello Vesta 2
time                 2.378 ms   (2.371 ms .. 2.398 ms)

benchmarking Group 1/Pasta base * -1
time                 92.48 μs   (92.35 μs .. 92.73 μs)

benchmarking Group 1/Pasta base * 2^200 - 1
time                 106.4 μs   (104.4 μs .. 109.1 μs)

benchmarking Group 1/Pasta base * 2^100 - 1
time                 31.51 μs   (31.45 μs .. 31.58 μs)

benchmarking Group 1/Vesta base * -1
time                 92.43 μs   (91.97 μs .. 92.79 μs)
-}

src/Crypto/ECC/Pasta.hs
{-|
Module      : Crypto.PastaCurves.Pasta (internal)
Description : Pasta-specific instantiation of parameterized curves and fields.
Copyright   : (c) Eric Schorn, 2022
Maintainer  : eric.schorn@nccgroup.com
Stability   : experimental
Portability : GHC
SPDX-License-Identifier: MIT

This internal module instantiates the specific curves and fields specific to Pasta. It
also includes `hashToPallas` and `hashToVesta` functionality (which in turn includes
two isogenous curves, mapping functionality, and coefficient vectors). The algorithms 
are NOT constant time; Safety and simplicity are the top priorities.
-}

{-# LANGUAGE DataKinds, NoImplicitPrelude, ScopedTypeVariables, Safe #-}

module Pasta (Fp, Fq, Num(..), Pallas, Vesta, Curve(..), CurvePt(..), Field(..), 
  hashToPallas, hashToVesta, rndPallas, rndVesta, pallasPrime, vestaPrime) where

import Prelude
import Curves (Curve(..), CurvePt(..), Point(..))
import Fields (Fz, Field(..))
import Data.ByteString.UTF8 (ByteString)
import System.Random (RandomGen)


-- | `Fp` is the field element used as a coordinate in the Pallas elliptic curve.
-- It is a type synonym for the internal `Fields.Fz` type, parameterized with the 
-- correct modulus. It is also typically used as a scalar to multiply a Vesta elliptic
-- curve point. Note that pointMul does not enforce specific scalar/point combinations.
type Fp  = Fz 0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001


-- | `Pallas` represents a point on the Pallas elliptic curve using `Fp` coordinates.
-- The curve was designed to have the some order as the `Fq` element\'s modulus. It is
-- a type synonym for the internal `Curves.Point` type, parameterized with the curve\s 
-- @a@ and @b@ values and the affine base point as @base_x@ and @base_y@. The underlying
-- point is of type @Point a b base_x base_y field@.
type Pallas = (Point 0 5 1 0x248b4a5cf5ed6c83ac20560f9c8711ab92e13d27d60fb1aa7f5db6c93512d546 Fp)


-- | `Fq` is the field element used as a coordinate in the Vesta elliptic curve.
-- It is a type synonym for the internal `Fields.Fz` type, parameterized with the 
-- correct modulus. It is also typically used as a scalar to multiply a Pallas elliptic 
-- curve point. Note that pointMul does not enforce specific scalar/point combinations.
type Fq = Fz 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001


-- | `Vesta` represents a point on the Vesta elliptic curve using `Fq` coordinates.
-- The curve was designed to have the some order as the `Fp` element\'s modulus.  It is
-- a type synonym for the internal `Curves.Point` type, parameterized with the curve\s 
-- @a@ and @b@ values and the affine base point as @base_x@ and @base_y@.  The underlying
-- point is of type @Point a b base_x base_y field@.
type Vesta  = (Point 0 5 1 0x26bc999156dd5194ec49b1c551768ab375785e7ce00906d13e0361674fd8959f Fq)


-- This is a curve that is isogenous to Pallas, but with a*b != 0; base point params are unused
type IsoPallas = (Point 0x18354a2eb0ea8c9c49be2d7258370742b74134581a27a59f92bb4b0b657a014b 1265 0 0 Fp)


-- This is a curve that is isogenous to Vesta, but with a*b != 0; base point params are unused
type IsoVesta = (Point 0x267f9b2ee592271a81639c4d96f787739673928c7d01b212c515ad7242eaa6b1 1265 0 0 Fq)


-- | The `hashToPallas` function takes an arbitrary `ByteString` and maps it to a valid 
-- point on the Pallas elliptic curve (of unknown relation to the base point).
hashToPallas :: ByteString -> Pallas
hashToPallas msg = result 
  where
    (fe0, fe1) = hash2Field msg "z.cash:test" "pallas" :: (Fp, Fp)
    q0 = mapToCurveSimpleSwu fe0 (fromInteger (-13)) :: IsoPallas  -- -13 is Pasta specific magic constant
    q1 = mapToCurveSimpleSwu fe1 (fromInteger (-13)) :: IsoPallas
    (Projective xp yp zp) = pointAdd q0 q1 :: IsoPallas
    x = xp * inv0 zp ;  y = yp * inv0 zp
    xTop = head isoPallasVecs * x ^ (3::Integer) + isoPallasVecs !! 1 * x ^ (2::Integer) + isoPallasVecs !! 2 * x + isoPallasVecs !! 3
    xBot = x ^ (2::Integer) + isoPallasVecs !! 4 * x + isoPallasVecs !! 5
    yTop = isoPallasVecs !! 6 * x ^ (3::Integer) + isoPallasVecs !! 7 * x ^ (2::Integer) + isoPallasVecs !! 8 * x + isoPallasVecs !! 9
    yBot = x ^ (3::Integer) + isoPallasVecs !! 10 * x ^ (2::Integer) + isoPallasVecs !! 11 * x + isoPallasVecs !! 12 
    proposed = Projective (xTop * inv0 xBot) (y * yTop * inv0 yBot) 1 :: Pallas
    result = if isOnCurve proposed then proposed else error "hashed to Pallas non-point"


-- | The `hashToVesta` function takes an arbitrary `ByteString` and maps it to a valid 
-- point on the Vesta elliptic curve (of unknown relation to the base point).
hashToVesta :: ByteString -> Vesta
hashToVesta msg = result 
  where
    (fe0, fe1) = hash2Field msg "z.cash:test" "vesta" :: (Fq, Fq)
    q0 = mapToCurveSimpleSwu fe0 (fromInteger (-13)) :: IsoVesta
    q1 = mapToCurveSimpleSwu fe1 (fromInteger (-13)) :: IsoVesta
    (Projective xp yp zp) = pointAdd q0 q1 :: IsoVesta
    x = xp * inv0 zp ;  y = yp * inv0 zp
    xTop = head isoVestaVecs * x ^ (3::Integer) + isoVestaVecs !! 1 * x ^ (2::Integer) + isoVestaVecs !! 2 * x + isoVestaVecs !! 3
    xBot = x ^ (2::Integer) + isoVestaVecs !! 4 * x + isoVestaVecs !! 5
    yTop = isoVestaVecs !! 6 * x ^ (3::Integer) + isoVestaVecs !! 7 * x ^ (2::Integer) + isoVestaVecs !! 8 * x + isoVestaVecs !! 9
    yBot = x ^ (3::Integer) + isoVestaVecs !! 10 * x ^ (2::Integer) + isoVestaVecs !! 11 * x + isoVestaVecs !! 12 
    proposed = Projective (xTop * inv0 xBot) (y * yTop * inv0 yBot) 1 :: Vesta
    result = if isOnCurve proposed then proposed else error "hashed to Vesta non-point"


-- | The `rndPallas` function returns a random Pallas point when given a StdGen.
rndPallas :: forall g. RandomGen g => g -> (g, Pallas)
rndPallas rndGen = hashToPallas . toBytesF <$> (rndF rndGen :: (g, Fq))


-- | The `rndVesta` function returns a random Vests point when given a StdGen.
rndVesta :: forall g. RandomGen g => g -> (g, Vesta)
rndVesta rndGen = hashToVesta . toBytesF <$> (rndF rndGen :: (g, Fp))


-- | The Pallas field modulus https://neuromancer.sk/std/other/Pallas
pallasPrime :: Integer
pallasPrime = 0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001


-- | The Vesta field modulus https://neuromancer.sk/std/other/Vesta
vestaPrime :: Integer
vestaPrime = 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001


-- Vectors to map isogenous curve (with a*b != 0) point to Pallas
-- See https://github.com/zcash/pasta_curves/blob/21fd9e2c1bbd2d049bfe95588d77cb884e9f93ab/src/curves.rs#L1017-L1096
isoPallasVecs :: [Fp]
isoPallasVecs = [
  0x0e38e38e38e38e38e38e38e38e38e38e4081775473d8375b775f6034aaaaaaab,
  0x3509afd51872d88e267c7ffa51cf412a0f93b82ee4b994958cf863b02814fb76,
  0x17329b9ec525375398c7d7ac3d98fd13380af066cfeb6d690eb64faef37ea4f7,
  0x1c71c71c71c71c71c71c71c71c71c71c8102eea8e7b06eb6eebec06955555580,
  0x1d572e7ddc099cff5a607fcce0494a799c434ac1c96b6980c47f2ab668bcd71f,
  0x325669becaecd5d11d13bf2a7f22b105b4abf9fb9a1fc81c2aa3af1eae5b6604,
  0x1a12f684bda12f684bda12f684bda12f7642b01ad461bad25ad985b5e38e38e4,
  0x1a84d7ea8c396c47133e3ffd28e7a09507c9dc17725cca4ac67c31d8140a7dbb,
  0x3fb98ff0d2ddcadd303216cce1db9ff11765e924f745937802e2be87d225b234,
  0x025ed097b425ed097b425ed097b425ed0ac03e8e134eb3e493e53ab371c71c4f,
  0x0c02c5bcca0e6b7f0790bfb3506defb65941a3a4a97aa1b35a28279b1d1b42ae,
  0x17033d3c60c68173573b3d7f7d681310d976bbfabbc5661d4d90ab820b12320a,
  0x40000000000000000000000000000000224698fc094cf91b992d30ecfffffde5]


-- Vectors to map isogenous curve (with a*b != 0) point to Vesta
-- Curve isogenous to Vesta, with a*b != 0
-- See https://github.com/zcash/pasta_curves/blob/21fd9e2c1bbd2d049bfe95588d77cb884e9f93ab/src/curves.rs#L1117-L1196
isoVestaVecs :: [Fq]
isoVestaVecs = [
  0x38e38e38e38e38e38e38e38e38e38e390205dd51cfa0961a43cd42c800000001,
  0x1d935247b4473d17acecf10f5f7c09a2216b8861ec72bd5d8b95c6aaf703bcc5,
  0x18760c7f7a9ad20ded7ee4a9cdf78f8fd59d03d23b39cb11aeac67bbeb586a3d,
  0x31c71c71c71c71c71c71c71c71c71c71e1c521a795ac8356fb539a6f0000002b,
  0x0a2de485568125d51454798a5b5c56b2a3ad678129b604d3b7284f7eaf21a2e9,
  0x14735171ee5427780c621de8b91c242a30cd6d53df49d235f169c187d2533465,
  0x12f684bda12f684bda12f684bda12f685601f4709a8adcb36bef1642aaaaaaab,
  0x2ec9a923da239e8bd6767887afbe04d121d910aefb03b31d8bee58e5fb81de63,
  0x19b0d87e16e2578866d1466e9de10e6497a3ca5c24e9ea634986913ab4443034,
  0x1ed097b425ed097b425ed097b425ed098bc32d36fb21a6a38f64842c55555533,
  0x2f44d6c801c1b8bf9e7eb64f890a820c06a767bfc35b5bac58dfecce86b2745e,
  0x3d59f455cafc7668252659ba2b546c7e926847fb9ddd76a1d43d449776f99d2f,
  0x40000000000000000000000000000000224698fc0994a8dd8c46eb20fffffde5]


src/Crypto/ECC/PastaCurves.hs
{-|
Module      : Crypto.PastaCurves.PastaCurves
Description : Provides the overall Pasta curve and field functionality.
Copyright   : (c) Eric Schorn, 2022
Maintainer  : eric.schorn@nccgroup.com
Stability   : experimental
Portability : GHC
SPDX-License-Identifier: MIT

This module provides the Pasta Curves consisting of: the `Pallas` curve and its `Fp` 
field element, the `Vesta` curve  and its `Fq` field element, and a variety of 
supporting functionality such as point/element arithmetic, serialization, and 
hash-to-curve. The algorithms are NOT constant time; Safety and simplicity are the top 
priorities.

\[
\text{Pallas: } y^2 = x^3 + 5 \text{ over } F_p(0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001)
\]
    
\[
\text{Vesta:  }  y^2 = x^3 + 5 \text{ over } F_q(0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001)
\]

The order of the Pallas curve is 0x40000000000000000000000000000000224698fc0994a8dd8c46eb2100000001.

The order of the Vesta curve is 0x40000000000000000000000000000000224698fc094cf91b992d30ed00000001.

The curves are designed such that the order of one matches the field characteristic of
the other. For a brief introduction, see the Zcash blog titled "The Pasta Curves for Halo 
2 and Beyond" at <https://electriccoin.co/blog/the-pasta-curves-for-halo-2-and-beyond/>.
The reference Rust implementation (which inspired this implementation) can be found at:
<https://github.com/zcash/pasta_curves>.

Example usage of this library implementation:

@
$ cabal repl

ghci> a = 9 :: Fp

ghci> a*a
0x0000000000000000000000000000000000000000000000000000000000000051

ghci> pointMul a base :: Vesta
Projective {_px = 0x3CDC6A090F2BB3B52714C083929B620FE24ADBCBBD420752108CD7C29E543E5E, 
            _py = 0x08795CD330B3CE5AA63BD2B18DE155AE3C96E8AF9DA2CC742C6BA1464E490161, 
            _pz = 0x1FA26F58F3A641ADFE81775D3D53378D6178B6CCBF14F9BD4AB5F10DEE28D878}
@

-}

{-# LANGUAGE DataKinds, NoImplicitPrelude, Safe #-}

module PastaCurves (Fp, Fq, Pallas, Vesta, CurvePt(..), Curve(..), hashToPallas, 
  hashToVesta, rndPallas, rndVesta, Field(..), pallasPrime, vestaPrime, exampleFp, 
  exampleFq, examplePallasPt, exampleVestaPt, Num(..)) where

import Pasta (Fp, Fq, Num(..), Pallas, Vesta, CurvePt(..), Curve(..), Field(..), 
  hashToPallas, hashToVesta, rndPallas, rndVesta, pallasPrime, vestaPrime)
import Data.ByteString.UTF8 (fromString)


-- | An example `Fp` element (8).
exampleFp :: Fp
exampleFp = 8


-- | An example `Fq` element (999).
exampleFq :: Fq
exampleFq = 999


-- | An example `Pallas` point generated by hashing a message.
examplePallasPt :: Pallas
examplePallasPt = hashToPallas (fromString "A message to hash")


-- | An example `Vesta` point (base * 8).
exampleVestaPt :: Vesta
exampleVestaPt = pointMul exampleFp base
